In [2]:
import cv2
import pandas as pd

import os
from glob import glob
from pathlib import Path

In [3]:
DATA_DIR = "C:\\Users\\Lenovo\\1navigation\\circles_2024_marks_surr\\cut"

In [5]:
PREPROC_DATA_DIR = 'downsamp_preprocessed_dataset'

# 1. open files
dataset_names = sorted(glob(os.path.join(DATA_DIR, '*.png')))
images = {}
for nam in dataset_names:
    print(nam)
    k = Path(nam).stem
    k = k.replace(' = ', '')
    k = k.replace(' ', '_')
    k = k.replace(',', '')
    im = cv2.imread(nam)
    images[k] = im


C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr3, center, r = 2.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr3, center, r = 6.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr3, walls, r = 2.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr3, walls, r = 6.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr5, center, r = 2.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr5, center, r = 6.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr5, walls, r = 2.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr5, walls, r = 6.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr6, center, r = 2.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr6, center, r = 6.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr6, walls, r = 2.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr6, walls, r = 6.png
C:\Users\Lenovo\1navigation\circles_2024_marks_surr\cut\tr

In [9]:
#count areas of different colors
colors_areas = pd.DataFrame()


for n, im in images.items():
    coords = []
    for i in range(len(im)):
        for j in range(len(im[i])):
            coords.append([i,j,im[i][j][2],im[i][j][1],im[i][j][0]])
    df = pd.DataFrame(coords, columns =['y', 'x', 'red', 'green', 'blue'])
    df['color'] = 'Not defined'
    df.loc[(df['green']>200)&(df['blue']<100)&(df['red']<100), 'color'] = 'green' 
    df.loc[(df['blue']>200)&(df['green']<100)&(df['red']<100), 'color'] = 'blue' 
    df.loc[(df['red']>200)&(df['green']>200)&(df['blue']>200), 'color'] = 'white'
    
    lildf = pd.DataFrame({'name_of_picture':[n,n,n,n], 'color':['green','blue','white','not defined'], \
                             'area_perc': [((len(df.loc[df['color']=='green']) / len(df)) * 100), (len(df.loc[df['color']=='blue']) / len(df) * 100),\
                                           ((len(df.loc[df['color']=='white']) / len(df))*100), (len(df.loc[df['color']=='Not defined']) / len(df)* 100)]})
    colors_areas = colors_areas.append(lildf)

print(colors_areas)



C:\Users\Lenovo\AppData\Local\Temp\ipykernel_13732\1658438400.py:19: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  colors_areas = colors_areas.append(lildf)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_13732\1658438400.py:19: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  colors_areas = colors_areas.append(lildf)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_13732\1658438400.py:19: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  colors_areas = colors_areas.append(lildf)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_13732\1658438400.py:19: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  colors_areas = colors_areas.append(lildf)
C:\Users\Lenovo\

   name_of_picture        color  area_perc
0    tr3_center_r2        green  35.999794
1    tr3_center_r2         blue   1.938253
2    tr3_center_r2        white  61.938066
3    tr3_center_r2  not defined   0.123887
0    tr3_center_r6        green  35.722737
..             ...          ...        ...
3    tr91_walls_r2  not defined   0.687224
0    tr91_walls_r6        green  64.974035
1    tr91_walls_r6         blue  32.757228
2    tr91_walls_r6        white   2.165441
3    tr91_walls_r6  not defined   0.103295

[80 rows x 3 columns]


C:\Users\Lenovo\AppData\Local\Temp\ipykernel_13732\1658438400.py:19: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  colors_areas = colors_areas.append(lildf)


In [10]:
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(colors_areas)

  name_of_picture        color  area_perc
0   tr3_center_r2        green     36.000
1   tr3_center_r2         blue      1.938
2   tr3_center_r2        white     61.938
3   tr3_center_r2  not defined      0.124
0   tr3_center_r6        green     35.723
1   tr3_center_r6         blue     13.838
2   tr3_center_r6        white     50.208
3   tr3_center_r6  not defined      0.232
0    tr3_walls_r2        green     64.834
1    tr3_walls_r2         blue      5.092
2    tr3_walls_r2        white     29.847
3    tr3_walls_r2  not defined      0.227
0    tr3_walls_r6        green     64.834
1    tr3_walls_r6         blue     16.105
2    tr3_walls_r6        white     18.930
3    tr3_walls_r6  not defined      0.132
0   tr5_center_r2        green     35.303
1   tr5_center_r2         blue      0.669
2   tr5_center_r2        white     63.979
3   tr5_center_r2  not defined      0.050
0   tr5_center_r6        green     35.723
1   tr5_center_r6         blue      8.085
2   tr5_center_r6        white    

In [12]:
colors_areas.to_csv('marks_areas_colors_areas.csv')

In [82]:
#count areas of different colors
for n, im in images.items():
    coords = []
    for i in range(len(im)):
        for j in range(len(im[i])):
            coords.append([i,j,im[i][j][2],im[i][j][1],im[i][j][0]])
    df = pd.DataFrame(coords, columns =['y', 'x', 'red', 'green', 'blue'])
    df['color'] = 'Not defined'
    df.loc[(df['green']>200)&(df['blue']<100)&(df['red']<100), 'color'] = 'green' 
    df.loc[(df['blue']>200)&(df['green']<100)&(df['red']<100), 'color'] = 'blue' 
    df.loc[(df['red']>200)&(df['green']>200)&(df['blue']>200), 'color'] = 'white'
    print(n, 'whole df, ', len(df))
    print(n, 'green,',(len(df.loc[df['color']=='green']) / len(df)) * 100)
    print(n, 'blue,',(len(df.loc[df['color']=='blue']) / len(df))* 100)
    print(n, 'white,',(len(df.loc[df['color']=='white']) / len(df))* 100)
    print(n, 'Not defined,',(len(df.loc[df['color']=='Not defined'])/ len(df))* 100)
    

tr3_center_r2 whole df,  213904
tr3_center_r2 green, 35.99979430024684
tr3_center_r2 blue, 1.938252674096791
tr3_center_r2 white, 61.93806567432119
tr3_center_r2 Not defined, 0.12388735133517839
tr3_center_r6 whole df,  212982
tr3_center_r6 green, 35.722737132715444
tr3_center_r6 blue, 13.837789108938784
tr3_center_r6 white, 50.2075292747744
tr3_center_r6 Not defined, 0.23194448357138162
tr3_walls_r2 whole df,  213443
tr3_walls_r2 green, 64.83370267471878
tr3_walls_r2 blue, 5.092226027557708
tr3_walls_r2 white, 29.846844356572948
tr3_walls_r2 Not defined, 0.2272269411505648
tr3_walls_r6 whole df,  213443
tr3_walls_r6 green, 64.83370267471878
tr3_walls_r6 blue, 16.105002272269413
tr3_walls_r6 white, 18.929643979891587
tr3_walls_r6 Not defined, 0.13165107312022414
tr5_center_r2 whole df,  211599
tr5_center_r2 green, 35.302624303517504
tr5_center_r2 blue, 0.6687177160572593
tr5_center_r2 white, 63.979035817749605
tr5_center_r2 Not defined, 0.04962216267562701
tr5_center_r6 whole df,  2129

tr3_center_r2 whole df,  213904
tr3_center_r2 green, 77005
tr3_center_r2 blue, 4146
tr3_center_r2 white, 132488
tr3_center_r2 Not defined, 265
tr3_center_r6 whole df,  212982
tr3_center_r6 green, 76083
tr3_center_r6 blue, 29472
tr3_center_r6 white, 106933
tr3_center_r6 Not defined, 494
tr3_walls_r2 whole df,  213443
tr3_walls_r2 green, 138383
tr3_walls_r2 blue, 10869
tr3_walls_r2 white, 63706
tr3_walls_r2 Not defined, 485
tr3_walls_r6 whole df,  213443
tr3_walls_r6 green, 138383
tr3_walls_r6 blue, 34375
tr3_walls_r6 white, 40404
tr3_walls_r6 Not defined, 281
tr5_center_r2 whole df,  211599
tr5_center_r2 green, 74700
tr5_center_r2 blue, 1415
tr5_center_r2 white, 135379
tr5_center_r2 Not defined, 105
tr5_center_r6 whole df,  212982
tr5_center_r6 green, 76083
tr5_center_r6 blue, 17219
tr5_center_r6 white, 119315
tr5_center_r6 Not defined, 365
tr5_walls_r2 whole df,  213443
tr5_walls_r2 green, 138383
tr5_walls_r2 blue, 15775
tr5_walls_r2 white, 58635
tr5_walls_r2 Not defined, 650
tr5_walls_r6 whole df,  212982
tr5_walls_r6 green, 138383
tr5_walls_r6 blue, 41685
tr5_walls_r6 white, 32557
tr5_walls_r6 Not defined, 357
tr6_center_r2 whole df,  214368
tr6_center_r2 green, 77005
tr6_center_r2 blue, 1820
tr6_center_r2 white, 134990
tr6_center_r2 Not defined, 553
tr6_center_r6 whole df,  213440
tr6_center_r6 green, 76541
tr6_center_r6 blue, 9002
tr6_center_r6 white, 127720
tr6_center_r6 Not defined, 177
tr6_walls_r2 whole df,  212060
tr6_walls_r2 green, 138383
tr6_walls_r2 blue, 8722
tr6_walls_r2 white, 64589
tr6_walls_r2 Not defined, 366
tr6_walls_r6 whole df,  213443
tr6_walls_r6 green, 138383
tr6_walls_r6 blue, 28070
tr6_walls_r6 white, 46704
tr6_walls_r6 Not defined, 286
tr7_center_r2 whole df,  213443
tr7_center_r2 green, 76544
tr7_center_r2 blue, 1354
tr7_center_r2 white, 135462
tr7_center_r2 Not defined, 83
tr7_center_r6 whole df,  213904
tr7_center_r6 green, 77005
tr7_center_r6 blue, 8280
tr7_center_r6 white, 128517
tr7_center_r6 Not defined, 102
tr7_walls_r2 whole df,  213904
tr7_walls_r2 green, 138383
tr7_walls_r2 blue, 10860
tr7_walls_r2 white, 64177
tr7_walls_r2 Not defined, 484
tr7_walls_r6 whole df,  213904
tr7_walls_r6 green, 138383
tr7_walls_r6 blue, 26424
tr7_walls_r6 white, 48966
tr7_walls_r6 Not defined, 131
tr91_center_r2 whole df,  213904
tr91_center_r2 green, 77005
tr91_center_r2 blue, 21991
tr91_center_r2 white, 113778
tr91_center_r2 Not defined, 1130
tr91_center_r6 whole df,  213443
tr91_center_r6 green, 76544
tr91_center_r6 blue, 84194
tr91_center_r6 white, 51959
tr91_center_r6 Not defined, 746
tr91_walls_r2 whole df,  213904
tr91_walls_r2 green, 138383
tr91_walls_r2 blue, 38337
tr91_walls_r2 white, 35714
tr91_walls_r2 Not defined, 1470
tr91_walls_r6 whole df,  212982
tr91_walls_r6 green, 138383
tr91_walls_r6 blue, 69767
tr91_walls_r6 white, 4612
tr91_walls_r6 Not defined, 220

#### draft

In [2]:
im3_2_c = cv2.imread('C:\\Users\\Lenovo\\1navigation\\circles_2024_marks_surr\\cut\\tr3, center, r = 2.png')

In [40]:
coords = []
for i in range(len(im3_2_c)):
    for j in range(len(im3_2_c[i])):
        coords.append([i,j,im3_2_c[i][j][2],im3_2_c[i][j][1],im3_2_c[i][j][0]])
df = pd.DataFrame(coords, columns =['y', 'x', 'red', 'green', 'blue'])

In [42]:
df['color'] = 'Not defined'
df.loc[(df['green']>200)&(df['blue']<100)&(df['red']<100), 'color'] = 'green' 
df.loc[(df['blue']>200)&(df['green']<100)&(df['red']<100), 'color'] = 'blue' 
df.loc[(df['red']>200)&(df['green']>200)&(df['blue']>200), 'color'] = 'white' 

In [48]:
df.loc[(df['x']==56) & (df['y']==414)]
#проверила что пиксели верно отображаются: https://pixspy.com/ 

y   x  red  green  blue color
192566  414  56   20     70   205  blue

In [50]:
print('green,',len(df.loc[df['color']=='green']))
print('blue,',len(df.loc[df['color']=='blue']))
print('white,',len(df.loc[df['color']=='white']))
print('Not defined,',len(df.loc[df['color']=='Not defined']))

green, 77005
blue, 4146
white, 132488
Not defined, 726
